python/uv/private/lock.bzl

```
@@ -12,114 +12,484 @@
+    srcs = ctx.files.srcs
+    python_version = full_version(
+        version = ctx.attr.python_version or DEFAULT_PYTHON_VERSION,
+        minor_mapping = MINOR_MAPPING,
+    )
+    output = ctx.actions.declare_file("{}.{}.out".format(
+        ctx.label.name,
+        python_version.replace(".", "_"),
+    ))
+
+    toolchain_info = ctx.toolchains[UV_TOOLCHAIN_TYPE]
+    uv = toolchain_info.uv_toolchain_info.uv[DefaultInfo].files_to_run.executable
+
+    args = _args(ctx)
+    args.add_all([
+        uv,
+        "pip",
+        "compile",
+        "--no-python-downloads",
+        "--no-cache",
+    ])
+    pkg = ctx.label.package
+    update_target = ctx.attr.update_target
+    args.add("--custom-compile-command", "bazel run //{}:{}".format(pkg, update_target))
+    if ctx.attr.generate_hashes:
+        args.add("--generate-hashes")
+    if not ctx.attr.strip_extras:
+        args.add("--no-strip-extras")
+    args.add_all(ctx.files.build_constraints, before_each = "--build-constraints")
+    args.add_all(ctx.files.constraints, before_each = "--constraints")
+    args.add_all(ctx.attr.args)
+
+    exec_tools = ctx.toolchains[EXEC_TOOLS_TOOLCHAIN_TYPE].exec_tools
+    runtime = exec_tools.exec_runtime.py3_runtime
+    python = runtime.interpreter or runtime.interpreter_path
+    python_files = runtime.files
+    args.add("--python", python)
+    args.add_all(srcs)
+
+    args.run_shell.add("--output-file", output)
+
+    # These arguments does not change behaviour, but it reduces the output from
+    # the command, which is especially verbose in stderr.
+    args.run_shell.add("--no-progress")
+    args.run_shell.add("--quiet")
 
-    Differences with the current {obj}`compile_pip_requirements` rule:
-    - This is implemented in shell and `uv`.
-    - This does not error out if the output file does not exist yet.
-    - Supports transitions out of the box.
-    - The execution of the lock file generation is happening inside of a build
-      action in a `genrule`.
+    if ctx.files.existing_output:
+        command = '{python} -c {python_cmd} && "$@"'.format(
+            python = getattr(python, "path", python),
+            python_cmd = shell.quote(
+                "from shutil import copy; copy(\"{src}\", \"{dst}\")".format(
+                    src = ctx.files.existing_output[0].path,
+                    dst = output.path,
+                ),
+            ),
+        )
+    else:
+        command = '"$@"'
+
+    srcs = srcs + ctx.files.build_constraints + ctx.files.constraints
+
+    ctx.actions.run_shell(
+        command = command,
+        inputs = srcs + ctx.files.existing_output,
+        mnemonic = "PyRequirementsLockUv",
+        outputs = [output],
+        arguments = [args.run_shell],
+        tools = [
+            uv,
+            python_files,
+        ],
+        progress_message = "Creating a requirements.txt with uv: //{}:{}".format(
+            ctx.label.package,
+            ctx.label.name,
+        ),
```

nit: Use %{label} instead

```suggestion
        progress_message = "Creating a requirements.txt with uv: %{label}",
```